providers/default_rhel.rb

```diff
@@ -0,0 +1,172 @@
+  end
+end
+
+def modprobe_file(action = :save)
+  # 'load' is the only special action otherwise 'save' and 'unload' should
+  # return the right path to create/remove the file
+  if action.to_s == 'load'
+    ::File.join(Chef::Config['file_cache_path'], new_resource.name + '.conf')
+  else
+    ::File.join(node['kernel_modules']['dirs']['modprobe.d'], new_resource.name + '.conf')
+  end
+end
+
+# Function to return the correct commnd for removing module
+def reload_cmd
+  # First unload the module nicely
+  unload_cmd = modprobe_bin + " -r #{new_resource.name}"
+  # If asked do it the strong way
+  unload_cmd + " || /sbin/rmmod -f #{new_resource.name}" if new_resource.force_reload
+  unload_cmd
+end
+
+def modprobe_bin(modprobe_bin = '/sbin/modprobe')
+  modprobe_bin
+end
+
+def resource_unloading(provider_action = :unload)
+  execute "Unloading module #{new_resource.name}" do
+    command reload_cmd
+    only_if "lsmod | grep -q #{new_resource.name}"
+
+    # In case of "save" or "load" we want to unload uniquely when the modules
+    # 'options' are changed only.
+    # For "unload" we unload no matter what
+    action provider_action.to_s == 'unload' ? :run : :nothing
+  end
+end
+
+def resource_load(provider_action = :save)
+  case node['platform_version']
+  when /^6/
+    # Script taken from '/etc/rc.sysinit#155' in CentOS 6
+    cmd = provider_action.to_s == 'save' ? "[ -x #{modload_file} ] && #{modload_file}" : "#{modprobe_bin} -C #{modprobe_file(:load)} #{new_resource.name}"
+
+    execute "Loading module #{new_resource.name}" do
+      command cmd
+      not_if "lsmod | grep -q #{new_resource.name}"
+      # No need to load the module if 'balcklist' is specify
+      not_if { ::KernelModules.modprobe_blacklisting?(new_resource.modprobe) }
+    end
+  when /^7/
+    service 'systemd-modules-load' do
+      subscribes :restart, 'template[modprobe_file(provider_action)]'
+      action [:enable, :start]
+    end
+
+  end
+end
+
+def resource_template_loading(provider_action = :save)
+  template modprobe_file(provider_action) do
+    source 'modprobe.conf.erb'
+    cookbook 'kernel-modules'
+    mode '0644'
+    variables(
+      name: new_resource.name,
+      config: new_resource.modprobe,
+    )
+    notifies :run, "execute[Unloading module #{new_resource.name}]", :immediately if new_resource.reload
+    only_if { new_resource.modprobe }
+
+    action provider_action.to_s == 'unload' ? :delete : :create
+  end
+end
+
+def failing_sanity_check
+  if error = sanity_check(new_resource.modprobe) # rubocop:disable Lint/AssignmentInCondition
+    raise "(Error) kernel-modules: the directive '#{error}' is not valid modprobe command"
+  end
+end
+
+# Make sure that the modprobe commands are valid
+def sanity_check(modprobe_options = {})
+  directives = %w(alias options install remove blacklist)
+  modprobe_options.keys.each do |c|
+    return c unless directives.include?(c.to_s)
+  end
+  nil
```

Comment:
I would replace the whole check with:

```
new_resource.modprobe.keys.all? {|cmd| directives.include? cmd }
```

But I'm wondering why we don't list each directive as a resource property since there are only 5 of them.